compiler/optimizer/OSRDefAnalysis.cpp

```diff
@@ -1378,6 +1408,82 @@ int32_t TR_OSRExceptionEdgeRemoval::perform()
          }
       }
 
+   // Under voluntary OSR, it is common for all paths to an OSR code block to contain
+   // a dead store for some of the prepareForOSR loads. As this is a cold path and
+   // OSR infrastructure is commonly ignored, the optimizer won't eliminate these loads.
+   //
+   // As this is cheap to do and will reduce the overhead of analysis like UseDef and
+   // CompactLocals, identify the loads that have dead stores on all paths
+   // and zero them.
+   //
+   static const char *disableDeadStoreCleanup = feGetEnv("TR_DisableDeadStoreCleanup");
+   if (!disableDeadStoreCleanup)
+      {
+      TR_BitVector alwaysDead(comp()->trMemory()->currentStackRegion());
+      TR_Stack<TR::Block *> osrBlocks(trMemory(), 8, false, stackAlloc);
+      _seenDeadStores = new (trStackMemory()) TR_BitVector(comp()->trMemory()->currentStackRegion());
+
+      // OSR induce and code blocks will contain dead stores for the next
+      // frame.
+      for (cfgNode = cfg->getFirstNode(); cfgNode; cfgNode = cfgNode->getNext())
+         {
+         TR::Block *block = toBlock(cfgNode);
+         if (!block->isOSRCodeBlock())
+            continue;
+
+         // Find the prepareForOSR call
+         TR::TreeTop *tt = block->getFirstRealTreeTop();
+         while (tt && !(tt->getNode()->getNumChildren() > 0 && tt->getNode()->getFirstChild()->getOpCode().isCall()))
```

Comment:
Shouldn't we check that the call is the call we expect by checking the symref for safety?

Reply:
A check has been added for the symref, with a more specific check for the treetop node.